File: runtime/tests/jvmtitests/src/com/ibm/jvmti/tests/followReferences/fr004.c
/*******************************************************************************
 * Copyright (c) 2001, 2018 IBM Corp. and others
 *
 * This program and the accompanying materials are made available under
 * the terms of the Eclipse Public License 2.0 which accompanies this
 * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
 * or the Apache License, Version 2.0 which accompanies this distribution and
 * is available at https://www.apache.org/licenses/LICENSE-2.0.
 *
 * This Source Code may also be made available under the following
 * Secondary Licenses when the conditions for such availability set
 * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
 * General Public License, version 2 with the GNU Classpath
 * Exception [1] and GNU General Public License, version 2 with the
 * OpenJDK Assembly Exception [2].
 *
 * [1] https://www.gnu.org/software/classpath/license.html
 * [2] http://openjdk.java.net/legal/assembly-exception.html
 *
 * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
 *******************************************************************************/
#include <string.h>

#include "fr.h"

static agentEnv * _agentEnv;

static int res_c1 = 0;
static int res_c2 = 0;
static int res_c3 = 0;
static int res_count = 0;

static jvmtiTestFollowRefsUserData userData;

jint JNICALL
fr004(agentEnv * env, char * args)
{
	JVMTI_ACCESS_FROM_AGENT(env);
	jvmtiCapabilities capabilities;
	jvmtiError err;                                

	if (!ensureVersion(env, JVMTI_VERSION_1_1)) {
		return JNI_ERR;
	}   
       
	_agentEnv = env;

	memset(&capabilities, 0, sizeof(jvmtiCapabilities));
	capabilities.can_tag_objects = 1;
	err = (*jvmti_env)->AddCapabilities(jvmti_env, &capabilities);
	if (err != JVMTI_ERROR_NONE) {
		error(env, err, "Failed to add capabilities");
		return JNI_ERR;
	}
	
	tagManager_initialize(env);
	
	return JNI_OK;
}


static jint JNICALL 
testFollowReferences_primitiveFieldCallback(jvmtiHeapReferenceKind reference_kind, const jvmtiHeapReferenceInfo* referrer_info, jlong class_tag, jlong* tag_ptr, jvalue value, jvmtiPrimitiveType value_type, void* user_data)
{
	
	if (*tag_ptr != 0L) {
		queueTag(*tag_ptr);
		queueTag(class_tag);
		printf("testFollowReferences_primitiveFieldCallback tag %llx:%llx\n", class_tag, *tag_ptr);
				
		if (value_type == JVMTI_PRIMITIVE_TYPE_INT) {
			if (value.i == 100)
				res_c1 = 100;
			if (value.i == 200)
				res_c2 = 200;
			if (value.i == 300)
				res_c3 = 300;			
		}
		
		res_count++;
		
	} else {
		if (class_tag != 0L) {
			printf("testFollowReferences_primitiveFieldCallback Class tag %llx\n", class_tag);
		}
	}
	
	return JVMTI_VISIT_OBJECTS;
}



jboolean JNICALL
Java_com_ibm_jvmti_tests_followReferences_fr004_followObjectPrimitiveFields(JNIEnv * jni_env, jclass klass, jobject initialObject, jclass filterClass)
{
	jvmtiError err;
	jvmtiHeapCallbacks callbacks;	
	JVMTI_ACCESS_FROM_AGENT(_agentEnv);
	jint errorCount;
	jint filterTag = JVMTI_HEAP_FILTER_UNTAGGED | JVMTI_HEAP_FILTER_CLASS_UNTAGGED;
	
	errorCount = getErrorCount();
	
	/* Test Follow References  Callback */
	memset(&callbacks, 0x00, sizeof(jvmtiHeapCallbacks));
	memset(&userData, 0x00, sizeof(jvmtiTestFollowRefsUserData));	
	callbacks.primitive_field_callback = testFollowReferences_primitiveFieldCallback;
	err = (*jvmti_env)->FollowReferences(jvmti_env, filterTag, filterClass, initialObject, &callbacks, (void *) &userData);
	if (err != JVMTI_ERROR_NONE) {
		error(_agentEnv, err, "FollowReferences failed");
		return JNI_FALSE;		
	}

	if (errorCount != getErrorCount()) {
		return JNI_FALSE;
	} 

	if (res_c1 != 100 || res_c2 != 200 || res_c3 != 300) {
		error(_agentEnv, JVMTI_ERROR_INTERNAL, "Received no primitive field callback for one or more of the expected fields");
		return JNI_FALSE;
	}
		
	if (res_count != 3) {
		error(_agentEnv, JVMTI_ERROR_INTERNAL, "Expected 3 primitive field callbacks, got %d", res_count);
		return JNI_FALSE;
	}
	
	return JNI_TRUE;
	 
}

File: test/functional/cmdLineTests/jvmtitests/src/com/ibm/jvmti/tests/followReferences/fr004.java
/*******************************************************************************
 * Copyright (c) 2001, 2018 IBM Corp. and others
 *
 * This program and the accompanying materials are made available under
 * the terms of the Eclipse Public License 2.0 which accompanies this
 * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
 * or the Apache License, Version 2.0 which accompanies this distribution and
 * is available at https://www.apache.org/licenses/LICENSE-2.0.
 *
 * This Source Code may also be made available under the following
 * Secondary Licenses when the conditions for such availability set
 * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
 * General Public License, version 2 with the GNU Classpath
 * Exception [1] and GNU General Public License, version 2 with the
 * OpenJDK Assembly Exception [2].
 *
 * [1] https://www.gnu.org/software/classpath/license.html
 * [2] http://openjdk.java.net/legal/assembly-exception.html
 *
 * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
 *******************************************************************************/
package com.ibm.jvmti.tests.followReferences;

public class fr004 
{

	public static native boolean followObjectPrimitiveFields(Object obj, Class filter);
	
	public boolean testFollowPrimitiveFields()
	{
		C1 c1 = new C1();
		C2 c2 = new C2();
		C3 c3 = new C3();
		
		TagManager.clearTags();
		TagManager.setObjectTag(c1, 0x1000L);
		TagManager.setObjectTag(c1.getClass(), 0x1001L);
		TagManager.setObjectTag(c2, 0x2000L);
		TagManager.setObjectTag(c2.getClass(), 0x2001L);		
		TagManager.setObjectTag(c3, 0x3000L);
		TagManager.setObjectTag(c3.getClass(), 0x3001L);
						
		boolean ret = followObjectPrimitiveFields(null, c3.getClass());
		if (ret == false)
			return ret;
			
		/* hold a ref to c3 such that we dont get GC'ed before heap iteration */	
		int foo = c3.c3_int;
		
		if (TagManager.isTagQueued(0x3001L) == false) {
			System.out.println("3001 missing");
			return false;
		}
		
		if (TagManager.isTagQueued(0x3000L) == false) {
			System.out.println("3000 missing");
			return false;
		}
		
		return ret;
	}
	
	public String helpFollowPrimitiveFields()
	{
		return "follow references from a specific object without any class filtering";
	}
	
}


class C1 {
	public int c1_int = 100;
	public static int c1_static_int = 101;
}

class C2 extends C1 {
	public int c2_int = 200;
	public static int c2_static_int = 201;
}

class C3 extends C2 {
	public int c3_int = 300;
	public static int c3_static_int = 301;
}
